File: Stage31/Java/Problem11049.java
package Stage31.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Problem11049 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        int[][] arr = new int[N][2];
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            arr[i][0] = Integer.parseInt(st.nextToken());
            arr[i][1] = Integer.parseInt(st.nextToken());
        }

        int[][] dp = new int[N][N];

        for (int len = 1; len < N; len++) {
            for (int i = 0; i + len < N; i++) {
                int j = i + len;
                dp[i][j] = Integer.MAX_VALUE;

                for (int k = i; k < j; k++) {
                    int cost = dp[i][k] + dp[k + 1][j]
                            + arr[i][0] * arr[k][1] * arr[j][1];

                    dp[i][j] = Math.min(dp[i][j], cost);
                }
            }
        }

        System.out.println(dp[0][N - 1]);
        br.close();
    }
}

File: Stage31/C/Problem11049.c
#include <stdio.h>
#include <limits.h>

int arr[501][2];
int dp[501][501];

int min(int a, int b) {
    return a < b ? a : b;
}

int main(void) {
    int N;
    scanf("%d", &N);

    for (int i = 0; i < N; i++) {
        scanf("%d %d", &arr[i][0], &arr[i][1]);
    }

    for (int i = 0; i < N; i++) {
        dp[i][i] = 0;
    }

    for (int len = 1; len < N; len++) {
        for (int i = 0; i + len < N; i++) {
            int j = i + len;
            dp[i][j] = INT_MAX;

            for (int k = i; k < j; k++) {
                int cost = dp[i][k] + dp[k + 1][j]
                         + arr[i][0] * arr[k][1] * arr[j][1];

                dp[i][j] = min(dp[i][j], cost);
            }
        }
    }

    printf("%d\n", dp[0][N - 1]);
    return 0;
}